python/tk_maya/panel_generation.py

```diff
@@ -0,0 +1,145 @@
+    """
+
+    # The imports are done here rather than at the module level to avoid spurious imports
+    # when this module is reloaded in the context of a workspace control UI script.
+    import maya.mel as mel
+    import maya.utils
+    import pymel.core as pm
+
+    # Create the Maya panel name.
+    maya_panel_id = "panel_%s" % panel_id
+
+    # When the Maya panel already exists, it can be delete safely since its embedded
+    # Shotgun app panel widget has already been reparented under Maya main window.
+    if pm.control(maya_panel_id, query=True, exists=True):
+        engine.log_debug("Deleting existing Maya panel %s." % maya_panel_id)
+        pm.deleteUI(maya_panel_id)
+
+    # Use the proper Maya panel docking method according to the Maya version.
+    if mel.eval("getApplicationVersionAsFloat()") < 2017:
+
+        # Create a new Maya window.
+        maya_window = pm.window()
+        engine.log_debug("Created Maya window %s." % maya_window)
+
+        # Add a layout to the Maya window.
+        maya_layout = pm.formLayout(parent=maya_window)
+        engine.log_debug("Created Maya layout %s." % maya_layout)
+
+        # Reparent the Shotgun app panel widget under the Maya window layout.
+        engine.log_debug("Reparenting Shotgun app panel widget %s under Maya layout %s." % (widget_id, maya_layout))
+        pm.control(widget_id, edit=True, parent=maya_layout)
+
+        # Keep the Shotgun app panel widget sides aligned with the Maya window layout sides.
+        pm.formLayout(maya_layout,
+                      edit=True,
+                      attachForm=[(widget_id, 'top', 1),
+                                  (widget_id, 'left', 1),
+                                  (widget_id, 'bottom', 1),
+                                  (widget_id, 'right', 1)]
+        )
+
+        # Dock the Maya window into a new tab of Maya Channel Box dock area.
+        engine.log_debug("Creating Maya panel %s." % maya_panel_id)
+        pm.dockControl(maya_panel_id, area="right", content=maya_window, label=title)
+
+        # Once Maya will have completed its UI update and be idle,
+        # raise (with "r=True") the new dock tab to the top.
+        maya.utils.executeDeferred("cmds.dockControl('%s', edit=True, r=True)" % maya_panel_id)
+
+    else:  # Maya 2017 and later
+
+        # Delete any default workspace control state that might have been automatically
+        # created by Maya when a previously existing Maya panel was closed and deleted.
+        if pm.workspaceControlState(maya_panel_id, exists=True):
+            engine.log_debug("Deleting existing Maya workspace panel state %s." % maya_panel_id)
+            pm.workspaceControlState(maya_panel_id, remove=True)
+
+        # Retrieve the Channel Box dock area, with error reporting turned on.
+        # This MEL function is declared in Maya startup script file UIComponents.mel.
+        dock_area = mel.eval('getUIComponentDockControl("Channel Box / Layer Editor", true)')
```

Comment:
This is the standard way to get a handle on the dock control. This MEL function is declared in Maya startup script file `UIComponents.mel`.

For now, when the Channel Box dock area cannot be found in the active Maya workspace, the Shotgun app panel widget is embedded into a floating workspace control window. This floating workspace control can then be docked into an existing dock area by the user.

I updated the code comments with this info.